datafusion/core/src/physical_plan/file_format/row_filter.rs

```diff
@@ -202,7 +202,18 @@ impl<'a> ExprRewriter for FilterCandidateBuilder<'a> {
     fn mutate(&mut self, expr: Expr) -> Result<Expr> {
         if let Expr::Column(Column { name, .. }) = &expr {
             if self.file_schema.field_with_name(name).is_err() {
-                return Ok(Expr::Literal(ScalarValue::Null));
+                // the column expr must be in the table schema
+                return match self.table_schema.field_with_name(name) {
```

Comment:
This would cause an error for projected columns right? Perhaps we can push this one level higher and rewrite binary expressions where the LHS is a column expression.

Reply:
Something like 
```
     fn mutate(&mut self, expr: Expr) -> Result<Expr> {
        match &expr {
            Expr::BinaryExpr { left: Column { name, .. }, op, right } => {
                if self.file_schema.field_with_name(name).is_err() {
                    return Ok(Expr::Literal(ScalarValue::Boolean(Some(false)));
                }
            }
            Expr::Column(Column { name, .. }) => {
                if self.file_schema.field_with_name(name).is_err() {
                    return Ok(Expr::Literal(ScalarValue::Null));
                }
            }
            _ => ()
        }

        Ok(expr)
    }
```

Reply:
If the expr matched the `Expr::Column`, it must be matched to a column in the table schema.
So we don't need to push it to higher level and rewrite the binary expression.
Just like the table schema is `c1,c2,c3`, and  file1 is `c1,c2`, the file2 is `c1,c3`, we use the filter c3 < 10.
Then we will produce the c3< 10 for file2, and NULL(with data type of c3) < 10.

@thinkharderdev

Reply:
> If the expr matched the Expr::Column, it must be matched to a column in the table schema.

This is not the case. The table schema is the schema prior to adding any projected columns. So if a predicate references a projected column then it won't be in the table schema

Reply:
> > If the expr matched the Expr::Column, it must be matched to a column in the table schema.
> 
> This is not the case. The table schema is the schema prior to adding any projected columns. So if a predicate references a projected column then it won't be in the table schema

Do you means that `select a+3 from table where a+3 > 4`? or  `select a from table abs(b) > 4`?

Reply:
In the case where the table has projected columns (like in `parquet_multiple_partitions` where we project columns from the hive partition directory structure) then the projected columns are not in the table schema. 

If the predicate contains any such projected columns then we ignore it anyway. So I think this change is fine but with the one adjustment that we don't return an error. So maybe

```
if let Ok(field) self.table_schema.field_with_name(name) {
    // return the null value corresponding to the data type
    let null_value = ScalarValue::try_from(field.data_type())?;
    return Ok(Expr::Literal(null_value));
}
```

In this case we just won't rewrite the expression but also `FilterCandidateBuilder.projected_columns` will be `true` so we won't crate a row filter predicate for the `Expr`.

Reply:
If there is a predicate with column references in it, in order to handle it reasonably in all cases, I suspect we need to know the type of that column. I definitely had issues in the past  trying to use `ScalarValue::Null` in a predicate rather than the properly typed null. 

I would suggest passing down the complete schema to this reader, or rewriting any references to projected columns to `NULL` higher up where we do have the complete schema and type information

That being said, keeping this PR's change minimal and returning a generic `DataType::Null` in the case where the column is present seems like a good way to proceed -- it unblocks @liukun4515  but doesn't break any other cases.

Reply:
tracked a issue for this https://github.com/apache/arrow-datafusion/issues/3492

I have some confused about how and when throw the error when building the row filter.